index.js

```diff
@@ -62,7 +62,15 @@ const match = (payload, pattern, callback) => {
            */
           } else if (element instanceof Object) {
             currentNode[key][index] = {}
-            tester(payload[key][index], element, currentNode[key][index])
+
+            payload[key].forEach((payloadItem, payloadIndex) => {
+              if(payloadItem instanceof Object) {
```

Comment:
I tried to fix it myself because I don't like petty code review comments but this pull request is coming from a fork and I don't have permission to edit it. Can you pass this code through the standardjs linter? I think there should be a space between the `if` and `(`

Otherwise, this looks good! I'm a bit worried that this code is starting to get out of hand :D

Reply:
Ah right! Linting added.

Reply:
>  I'm a bit worried that this code is starting to get out of hand :D
Indeed! But I think we can document the overall algorithm a bit more and it should be fine, as long as we have enough tests and building on top.